### ISSUE TITLE
[RFC] Edge sampler with excluding edges adjacent to low-degree nodes

### ISSUE THREAD

**Issue by jwzhi** [dmlc/dgl#5698]
# [RFC] Edge sampler with excluding edges adjacent to low-degree nodes
## 🚀 Feature
This issue is for extending the function of excluding the training edges in the edge sampler used in the link prediction training. 
We develop a new edge prediction sampler `EdgePredictionSamplerwithDegree` based on `dgl.dataloading.as_edge_prediction_sampler`.
Besides the function to exclude all the training edges in the original sampler, our new edge sampler can exclude the training edges only adjacent to the low-degree nodes. 
Experimental Implementation is here.
## Motivation
For the GNN training on link prediction task, the existence of training target edges during message passing leads to the overfitting issue.
When this edge is not excluded from the training graph, 
GNNs would use the message passing graph 
and look up the existence of training edges in the message passing graph instead of learning to predict it based on the graph structure and node features.

In the test process, GNNs may still directly search from the message passing graph to determine the test edge existence. If test edges are excluded from the graph, it will cause influence the test performance and lead to the overfitting issue.

The intuitive solution to avoid the training overfitting issue is to exclude all the training edges which has been implemented in `dgl.dataloading.as_edge_prediction_sampler`, but this exposes several challenges.
1. Excluding all target edges cannot be done in full batch training setting, since in full batch training, excluding all training edges is equivalent to excluding all edges in the graph
2. We found that in the mini-batch setting, it causes too much distraction for mini-batch graphs.
3. Setting the batch size to be 1 can solve the structure distraction problem, the mini-batch message passing graph will be too small. As a result, the training will be unstable and computationally inefficient.

An Example of 2-hop message passing graph for a mini-batch of size 4 is shown below:

![Computation_graph](https://jwzhi.github.io/images/projects/spottarget-rfc.png)

The red lines are training edges and the black lines are other message passing edges induced by the edges. 
As shown on the left, if all training edges are excluded during training, it results in three disconnected components 
and one of them is an isolated node. But if only training edges incident to low-degree nodes are excluded, as shown on the right, the train graph is still connected. 
There is not much distraction to the graph structure while taking the benefit of avoiding the overfitting issue.

Developing an edge sampler which can provide the users to only exclude a proportion of training edges can benefit the overall training process. 
Our experimental results below can also our edge sampler's superiority over the existing `dgl.dataloading.as_edge_prediction_sampler`.
## Alternatives
There is currently no alternatives in DGL to only removing the training edges adjacent to low-degree nodes outlined above.
## Pitch
The user experience will look like the following:
```python
sampler = NeighborSampler([15, 10, 5], prefetch_node_feats=['feat'])
exclude_target_degree = ... 
# -1 means remove all training edges
# 0 means do not remove training edges 
# positive numbers means remove edges incident to the nodes with degrees less than exclude_target_degree
if exclude_target_degree == -1:
    edge_sampler = as_edge_prediction_sampler(sampler, exclude="reverse_id", reverse_eids=reverse_eids, 
                                            negative_sampler=negative_sampler.Uniform(1))
elif exclude_target_degree > 0:
    edge_sampler = EdgePredictionSamplerwithDegree(sampler, exclude="reverse_id", reverse_eids=reverse_eids,
                                              negative_sampler=negative_sampler.Uniform(1),
                                              degree_threshold=exclude_target_degree)
elif exclude_target_degree == 0:
    edge_sampler = as_edge_prediction_sampler(sampler, reverse_eids=reverse_eids, 
                                            negative_sampler=negative_sampler.Uniform(1))
```
If users do not exclude training edges or exclude all training edges, they can use the original `as_edge_prediction_sampler`. 
If users need to remove edges incident to low-degree nodes, they can use our `EdgePredictionSamplerwithDegree`.

We achieve the function of excluding a proportion of training edges by only applying `dgl.dataloading.base._find_exclude_eids` function to the edges incident to low-degree nodes. 
## API and usage scenarios
Our edge sampler `EdgePredictionSamplerwithDegree` inherits from `dgl.dataloading.base.EdgePredictionSampler`. 
We add a new argument `degree_threshold` with the default value 10. 
The sampler will automatically remove the training edges incident to at least one low-degree node lower than the `degree_threshold`.
The other functions of `EdgePredictionSamplerwithDegree` are the same as `dgl.dataloading.as_edge_prediction_sampler`, 
so the usage scenarios of two edge samplers are almost the same. 
If the users need to exclude the edges incident to low-degree nodes, our new edge sampler can directly replace `as_edge_prediction_sampler` by adding the argument `degree_threshold`. 
 

## Experimental results
We select four common GNN models to validate our framework. SAGE, GCN and GATv2 are GAE-based models. 
SEAL is a subgraph-based model that extracts an enclosing subgraph for each target edge and then predicts the link likelihood based on the subgraph's embeddings. 

Besides our edge sampler which only excludes training edges adjacent to low-degree nodes (`ExcludeLow`), we use `ExcludeNone` and `ExcludeAll` as our baselines. 
`ExcludeNone` does not exclude any training edges, and `ExcludeAll` exclude all training edges.

Throughout the experiments, we use the numbers of training batch size 
for `Ogbl-collab`, `Ogbl-citation2`, and `USAir` is set to 256, 1024, and 64, respectively.
 
For SAGE models, the learning rate is 5e-4, the dimension of hidden layers is 256, and the number of hidden layers is 3. For GCN, the number of hidden layers is 1 and others remain the same as the SAGE. 
For GATv2, the number of heads is 8, the learning rate is 5e-4, the dimension of hidden layers is 64. 
For SEAL, the learning rate is 1e-4, the number of hops is 1, the training percent is 15%, 2% and 100% and 
the dimension of hidden layers is is 256, 256 and 32 for `Ogbl-collab`, `Ogbl-citation2`, and `USAir` respectively.

We used a single Nvidia A40 GPU to train the model and repeat our experiments with three different random seeds.

| Model | ExcludeNone      |     ExcludeAll    |    ExcludeLow    |
|-------|----------------------|:-----------------:|:----------------:|
|       |  Ogbl-Collab (H@50)  |                   |                  |
| SAGE  | 48.57 ± 0.74     |  45.82 ± 0.41 | 49.00 ± 0.65 |
| GCN   | 43.03 ± 0.50     |  37.75 ± 1.42 | 39.58 ± 1.06 |
| GATv2 | 45.61 ± 0.85     |  45.71 ± 0.87 | 45.46 ± 0.19 |
| SEAL  | 61.27 ± 0.28     |  64.11 ± 0.30 | 64.57 ± 0.30 |
|       | Ogbl-Citation2 (MRR) |                   |                  |
| SAGE  | 82.06 ± 0.06     |  81.47 ± 0.17 | 82.18 ± 0.18 |
| GCN   | 79.70 ± 0.25     |  79.06 ± 0.30 | 79.88 ± 0.14 |
| GATv2 | OOM                  |        OOM        |        OOM       |
| SEAL  | 86.75 ± 0.20     |  86.74 ± 0.23 | 86.93 ± 0.55 |
|       |      USAir (AUC)     |                   |                  |
| SAGE  | 95.97 ± 0.17     |  95.71 ± 0.12 | 96.19 ± 0.53 |
| GCN   | 94.00 ± 0.14     |  94.09 ± 0.11 | 94.28 ± 0.15 |
| GATv2 | 95.05 ± 0.66     | 95.66  ± 0.24 | 95.87 ± 0.46 |
| SEAL  | 95.36 ± 0.24     |  95.94 ± 0.04 | 96.39 ± 0.09 |
| Rank  | 2.27     |  2.45 | 1.27 |

The results indicate that `ExcludeLow` can achieve the best average results
and the lowest rank compared with `ExcludeNone` (not excluding training edges) and `ExcludeAll` (excluding all training edges).

**Comment by BarclayII** [comment]
If we only exclude a part of the training edges, do we compute losses only on those edges?  If we also compute losses on the edges not excluded during sampling, then would it cause label leakage during training?

**Comment by jwzhi** [comment]
Thanks for your comment. We do calculate the loss for the target edges that are not excluded, so these edges still suffer from the overfitting and distribution shift issues. However, since the target edges that we don't exclude are high-degree edges, the negative effects are diluted throughout the message passing process, and thus we found that excluding low-degree edges only achieves the best trade-off between avoiding overfitting and distribution shift issues as well as avoiding mini-batch graph structure corruption. Our paper will be put on Arxiv as soon as the internal review process is finished. 

Let me know if you have any questions on this.

**Comment by BarclayII** [comment]
My feeling towards this RFC is that this could be achieved by specifying a customized edge exclusion strategy in `as_edge_prediction_sampler`: it accepts an `exclude` argument which could be a user-defined function.  Does it achieve your goal?  Or are you proposing to add this functionality into what DGL official supports?

If the latter, I imagine that it's best to have an edge exclusion callable, something like a `dgl.dataloading.edge_exclusion.SpotTarget` class that has a `__call__` method:
```
class SpotTarget(object):
    def __init__(self, ...):
        ...
    def __call__(self, edge_minibatch):
        # return the edges to exclude given the minibatch
```

Then we can pass an instance into `as_edge_prediction_sampler`:
```
sampler = dgl.dataloading.as_edge_prediction_sampler(
    sampler,
    exclude=dgl.dataloading.edge_exclusion.SpotTarget(...),
)
```

What do you think?

cc @frozenbugs

**Comment by frozenbugs** [comment]
Yes, we can open a API to support python level UDF for edge filter, such that ppl can hack quickly for experiment.
If we believe this is a solid option, it is better to have C++ implementation.

@jwzhi Do you prefer to code in python or c++?

Just to confirm, when you mention high-degree / low-degree edges, are they in-degree or out-degree or sum of both?

For the actual implementation, we are working on refactoring the code for subgraph sampling, we can guide you towards the new implementation instead of just implementing it in existing dgl.dataloading.as_edge_prediction_sampler, which deemed to be deprecated in the future.

**Comment by BarclayII** [comment]
> Yes, we can open a API to support python level UDF for edge filter, such that ppl can hack quickly for experiment.

I think `as_edge_prediction_sampler` already has Python UDF support.

**Comment by jwzhi** [comment]
Hi, sorry for the late response. I was trying to catch up a paper deadline previously.  And I just found that I forgot to attach our code in the RFC request. 

 I prefer codes in Python. 

In our experiments, since all graphs are treated as undirected graph, we add the reverse edges into the graph and only take the in_degree of each node, see low. That's equivalent of taking the in_degree and out_degree for an undirected graph without reverse edges. I didn't do experiments on directed graphs. 

The SpotTarget code for low degree are shown below. Can you also point me towards the subgraph sampling code that you are trying to integrate to? And I can take a look at how to put this into the new implementation. 

```
from dgl.dataloading.base import EdgePredictionSampler, _find_exclude_eids
from collections.abc import Mapping
from dgl.base import EID, NID
from dgl.utils import recursive_apply
import dgl
import torch


def find_exclude_eids(g, seed_edges, exclude, reverse_eids=None, reverse_etypes=None,
                      output_device=None, degree_threshold=10):
    """Find all edge IDs to exclude according to :attr:`exclude_mode`.
    Parameters
    ----------
    g : DGLGraph
        The graph.
    exclude_mode : str, optional
        Can be either of the following,
        None (default)
            Does not exclude any edge.
        'self'
            Exclude the given edges themselves but nothing else.
        'reverse_id'
            Exclude all edges specified in ``eids``, as well as their reverse edges
            of the same edge type.
            The mapping from each edge ID to its reverse edge ID is specified in
            the keyword argument ``reverse_eid_map``.
            This mode assumes that the reverse of an edge with ID ``e`` and type
            ``etype`` will have ID ``reverse_eid_map[e]`` and type ``etype``.
        'reverse_types'
            Exclude all edges specified in ``eids``, as well as their reverse
            edges of the corresponding edge types.
            The mapping from each edge type to its reverse edge type is specified
            in the keyword argument ``reverse_etype_map``.
            This mode assumes that the reverse of an edge with ID ``e`` and type ``etype``
            will have ID ``e`` and type ``reverse_etype_map[etype]``.
        callable
            Any function that takes in a single argument :attr:`seed_edges` and returns
            a tensor or dict of tensors.
    eids : Tensor or dict[etype, Tensor]
        The edge IDs.
    reverse_eids : Tensor or dict[etype, Tensor]
        The mapping from edge ID to its reverse edge ID.
    reverse_etypes : dict[etype, etype]
        The mapping from edge etype to its reverse edge type.
    output_device : device
        The device of the output edge IDs.
    """
    # edges_need_to_exclude = TODO
    src, dst = g.find_edges(seed_edges)
    head_degree = g.in_degrees(src)
    tail_degree = g.in_degrees(dst)
    degree = torch.min(head_degree, tail_degree)
    degree_mask = degree < degree_threshold
    edges_need_to_exclude = seed_edges[degree_mask]
    exclude_eids = _find_exclude_eids(
        g,
        exclude,
        edges_need_to_exclude,
        reverse_eid_map=reverse_eids,
        reverse_etype_map=reverse_etypes)
    if exclude_eids is not None and output_device is not None:
        exclude_eids = recursive_apply(exclude_eids, lambda x: F.copy_to(x, output_device))
    return exclude_eids


class EdgePredictionSamplerwithDegree(EdgePredictionSampler):
    """Sampler class that builds upon EdgePredictionSampler
    The exlucde train target is only done on edges with a degree < 10
    TODO: Change the fixed degree to user-defined args

    ------------------------------
    Need to call this directly in the code instead of calling as_edge_prediction_sampler
    """

    def __init__(self, sampler, exclude=None, reverse_eids=None,
                 reverse_etypes=None, negative_sampler=None, prefetch_labels=None, degree_threshold=10):
        super().__init__(sampler, exclude, reverse_eids, reverse_etypes, negative_sampler, prefetch_labels)
        self.degree_threshold = degree_threshold

    def sample(self, g, seed_edges):  # pylint: disable=arguments-differ
        """Samples a list of blocks, as well as a subgraph containing the sampled
        edges from the original graph.
        If :attr:`negative_sampler` is given, also returns another graph containing the
        negative pairs as edges.
        """
        if isinstance(seed_edges, Mapping):
            seed_edges = {g.to_canonical_etype(k): v for k, v in seed_edges.items()}
        exclude = self.exclude
        pair_graph = g.edge_subgraph(
            seed_edges, relabel_nodes=False, output_device=self.output_device)
        eids = pair_graph.edata[EID]

        if self.negative_sampler is not None:
            neg_graph = self._build_neg_graph(g, seed_edges)
            pair_graph, neg_graph = dgl.compact_graphs([pair_graph, neg_graph])
        else:
            pair_graph = dgl.compact_graphs(pair_graph)

        pair_graph.edata[EID] = eids
        seed_nodes = pair_graph.ndata[NID]

        exclude_eids = find_exclude_eids(
            g, seed_edges, exclude, self.reverse_eids, self.reverse_etypes,
            self.output_device, self.degree_threshold)

        input_nodes, _, blocks = self.sampler.sample(g, seed_nodes, exclude_eids)

        if self.negative_sampler is None:
            return self.assign_lazy_features((input_nodes, pair_graph, blocks))
        else:
            return self.assign_lazy_features((input_nodes, pair_graph, neg_graph, blocks))
```

**Comment by BarclayII** [comment]
I think the only difference between naive exclude_edges and your SpotTarget is that your SpotTarget only exclude the seed edges with low-degree incident nodes.  I guess it's better to have it like a callable to avoid overriding `EdgePredictionSampler`.

Namely, I imagine that you could do something like (maybe you need some modifications):
```
class SpotTarget(object):
    def __init__(self, g, degree_threshold=10, ...):
        self.g = g
        self.degree_threshold = degree_threshold

    def __call__(self, edge_minibatch):
        g = self.g
        src, dst = g.find_edges(seed_edges)
        head_degree = g.in_degrees(src)
        tail_degree = g.in_degrees(dst)
        degree = torch.min(head_degree, tail_degree)
        degree_mask = degree < self.degree_threshold
        edges_need_to_exclude = seed_edges[degree_mask]
        return find_exclude_eids(g, edges_need_to_exclude, ...)
```
Then users could call it like:
```
sampler = dgl.dataloading.as_edge_prediction_sampler(
    sampler,
    exclude=dgl.dataloading.edge_exclusion.SpotTarget(g),
)
```

**Comment by jwzhi** [comment]
Oh I see. Sounds good! I will update the code and post when I finish. Thanks!

**Comment by BarclayII** [comment]
You could also directly make a PR.  The relevant code is in `dgl/dataloading/base.py`.

@frozenbugs Did our refactor go to edge exclusion in edge dataloader?  If not, I guess we can still have the change in `dgl/dataloader/base.py`, and do the refactor together?

**Comment by frozenbugs** [comment]
Yes, feel free to make the change in existing dataloader, we can figure out the future work of migrate the work to new design later.

**Comment by jwzhi** [comment]
Hi, I opened a pull request for this, see here https://github.com/dmlc/dgl/pull/5893. 

The issue I am having is, it has a CI test error, as reported by dgl-bot. But I didn't understand which part went wrong, according to the report. It would be great help if you can give me some ideas about this. 

Thanks,
Jing

**Comment by BarclayII** [comment]
Merged by #5893
